smartsim/entity/strategies.py

```diff
@@ -25,44 +25,97 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # Generation Strategies
+
+from __future__ import annotations
+
+import functools
+import itertools
 import random
 import typing as t
-from itertools import product
+
+from smartsim.error import errors
+
+TPermutationStrategy = t.Callable[
+    [t.Mapping[str, t.Sequence[str]], int], list[dict[str, str]]
+]
+
+_REGISTERED_STRATEGIES: t.Final[dict[str, TPermutationStrategy]] = {}
+
+
+def _register(name: str) -> t.Callable[
```

Comment:
Do you mind writing docstrings for all functions youve added in this file?

Reply:
Can do!